docs: new api (#472)
libp2p/js-libp2p: API.md

# API

> * [`create`](#create)

Reviewer: This is a static function but the rest are instance methods however they appear the same here. Might be a good idea to differentiate :)

* [`start`](#start)
* [`stop`](#stop)
* [`dial`](#dial)
* [`dialProtocol`](#dialProtocol)
* [`hangUp`](#hangUp)
* [`handle`](#handle)
* [`unhandle`](#unhandle)
* [`peerRouting.findPeer`](#peerRouting.findPeer)
* [`contentRouting.findProviders`](#contentRouting.findProviders)
* [`contentRouting.provide`](#contentRouting.provide)

> * [`dht.put`](#dht.put)

Reviewer 2: Suggestion: Avoid exposing the DHT methods as they are just methods from one impl of Peer Routing and Content Routing and instead stick with a clean Peer Routing & Content Routing interfaces.

Author: Shouldn't we make that clear instead?

As there are projects, including `IPFS` that use the `dht` API directly [interface-js-ipfs-core/SPEC/DHT.md](https://github.com/ipfs/interface-js-ipfs-core/blob/master/SPEC/DHT.md), I think we should make your point clear in the docs, but also have the `dht` available

Reviewer 2: A way to declare that the API is internal is by putting a `_` before the name (e.g. `_dht`), this way, folks are less tempted to use it.

* [`dht.provide`](#dht.provide)
* [`dht.findProviders`](#dht.findProviders)
* [`dht.getClosestPeers`](#dht.getClosestPeers)
* [`dht.getPublicKey`](#dht.getPublicKey)
* [`pubsub.getPeersSubscribed`](#pubsub.getPeersSubscribed)
* [`pubsub.getTopics`](#pubsub.getTopics)
* [`pubsub.publish`](#pubsub.publish)
* [`pubsub.subscribe`](#pubsub.subscribe)

## create

Creates an instance of Libp2p.

### `create(options)`

#### Parameters

| Name | Type | Description |
|------|------|-------------|
| options | `Object` | libp2p options |

#### Returns

| Type | Description |
|------|-------------|
| `Promise<Libp2p>` | Promise resolves with the Libp2p instance |

#### Example

```js
const { create } = require('libp2p')

// specify options
const options = {}

// create libp2p
const libp2p = await create(options)
```

Note: It is important pointing out that with `create`, the `PeerInfo` option is not required and will be generated if it is not provided.

As an alternative, it is possible to create a Libp2p instance with the constructor:

#### Example

```js
const Libp2p = require('libp2p')

// specify options
const options = {}

// create libp2p
const libp2p = new Libp2p (options)

> ```

Reviewer 2: js-ipfs had multiple rounds of iteration till it got its .create constructor https://github.com/ipfs/js-ipfs#ipfs-constructor, @alanshaw can shine some light here on why that was important. I would vote to have the same way of spawning a node just for consistency.

Author: I agree with the `create`. In `js-libp2p` we currently have `createLibp2p` and are moving into `create` now, same way as IPFS does.

I think this new API is consistent with `js-ipfs`. Same way, we also support using the `constructor`

Required keys in the `options` object:

- `peerInfo`: instance of [PeerInfo][] that contains the [PeerId][], Keys and [multiaddrs][multiaddr] of the libp2p Node.
- `modules.transport`: An array that must include at least 1 transport, such as `libp2p-tcp`.

## start

Starts the libp2p node.

### `libp2p.start()`

#### Returns

| Type | Description |
|------|-------------|
| `Promise` | Promise resolves when the node is ready |

#### Example

```js
const { create } = require('libp2p')

// ...

const libp2p = await create(options)

// start libp2p
await libp2p.start()
```

## stop

Stops the libp2p node.

### `libp2p.stop()`

#### Returns

| Type | Description |
|------|-------------|
| `Promise` | Promise resolves when the node is fully stopped |

#### Example

```js
const { create } = require('libp2p')

// ...

> const libp2p = await create(options)

Reviewer: A matter of preference but IMHO calling create on a default export is clearer in examples (and in app code):

```suggestion
const libp2p = await Libp2p.create(options)
```

// ...

// stop libp2p
await libp2p.stop()
```

## dial

Dials to another peer in the network, establishes the connection.

### `dial(peer, options)`

#### Parameters

| Name | Type | Description |
|------|------|-------------|
| peer | [PeerInfo](https://github.com/libp2p/js-peer-info), [PeerId](https://github.com/libp2p/js-peer-id), [multiaddr](https://github.com/multiformats/js-multiaddr), `string` | peer to dial |
| [options] | `Object` | dial options |

> | [options.abort] | `AbortSignal` | abort signal |

Reviewer: FYI these are named `options.signal` in a bunch of other places...including the `fetch` API. Personally I'd keep it consistent and familiar.

# API

* [Static Functions](#static-functions)
  * [`create`](#create)
* [Instance Methods](#instance-methods)
  * [`start`](#start)
  * [`stop`](#stop)
  * [`dial`](#dial)
  * [`dialProtocol`](#dialProtocol)
  * [`hangUp`](#hangUp)
  * [`handle`](#handle)
  * [`unhandle`](#unhandle)
  * [`peerRouting.findPeer`](#peerRouting.findPeer)
  * [`contentRouting.findProviders`](#contentRouting.findProviders)
  * [`contentRouting.provide`](#contentRouting.provide)
  * [`_dht.put`](#_dht.put)
  * [`_dht.get`](#_dht.get)

>   * [`_dht.getMany`](#_dht.getMany)

Reviewer: If these are not for public consumption they should not be included in the docs.

Author: I understand that! 

However, taking into account that project such as `js-ipfs` use these API methods, I also don't like that we do not document something that will probably be widely used (also discussed previously on https://github.com/libp2p/js-libp2p/pull/472#discussion_r345011828)

Author: I am updating to `contentRouting`

  * [`pubsub.getPeersSubscribed`](#pubsub.getPeersSubscribed)
  * [`pubsub.getTopics`](#pubsub.getTopics)
  * [`pubsub.publish`](#pubsub.publish)
  * [`pubsub.subscribe`](#pubsub.subscribe)

## Static Functions

### create

Creates an instance of Libp2p.

`create(options)`

#### Parameters

| Name | Type | Description |
|------|------|-------------|

> | options | `Object` | libp2p options |

Reviewer: What are the options?

Author: I would like to have a specific MD doc for the options/libp2p configuration, where this would point out. As there are so many options and instructions for configuration, I think it will be easier for people reading the docs to digest a lighter API, while being able to consult the options when needed. What do you think?

Author: I will create that in a new PR next week

Author: So, after writting that in #495 , I think I will document the first level properties of the options object here and point to the other doc

#### Returns

| Type | Description |
|------|-------------|
| `Promise<Libp2p>` | Promise resolves with the Libp2p instance |

#### Example

```js
const Libp2p = require('libp2p')

// specify options
const options = {}

// create libp2p
const libp2p = await Libp2p.create(options)
```

> Note: It is important pointing out that with `create`, the `PeerInfo` option is not required and will be generated if it is not provided.

Reviewer: ```suggestion
Note: The `PeerInfo` option is not required and will be generated if it is not provided.
```

> As an alternative, it is possible to create a Libp2p instance with the constructor:

Reviewer: In js-ipfs we did this:

<img width="460" alt="Screenshot 2019-11-29 at 15 29 49" src="https://user-images.githubusercontent.com/152863/69878567-39704780-12bd-11ea-9269-6050854c54b0.png">

...then clicking to expand:

<img width="905" alt="Screenshot 2019-11-29 at 15 30 10" src="https://user-images.githubusercontent.com/152863/69878582-42f9af80-12bd-11ea-92a6-895eb5eedf3b.png">

This helps to reduce the amount of info the reader has to read and understand (whilst still keeping it available) for methods that are deprecated or not recommended.

Author: cool, I will copy this approach! Thanks 🙏

#### Example

```js
const Libp2p = require('libp2p')

// specify options
const options = {}

// create libp2p

> const libp2p = new Libp2p (options)

Reviewer: ```suggestion
const libp2p = new Libp2p(options)
```

```

> Required keys in the `options` object:

Reviewer: You just said above that `peerInfo` was not required.

Author: `peerInfo` is not required for the `create`. With the alternative approach,i.e. the `constructor` it is required. This logic is prior do this refactor, if you want to discuss this, maybe we can open an issue for it?

Reviewer: Ok that makes sense.

- `peerInfo`: instance of [PeerInfo][] that contains the [PeerId][], Keys and [multiaddrs][multiaddr] of the libp2p Node.

> - `modules.transport`: An array that must include at least 1 transport, such as `libp2p-tcp`.

Reviewer: ```suggestion
- `modules.transport`: An array that must include at least 1 transport, such as [`libp2p-tcp`](https://github.com/libp2p/js-libp2p-tcp).
```

> - `modules.transport`: An array that must include at least 1 transport, such as `libp2p-tcp`.

Reviewer: There are many more options right? Are you planning on documenting them or at least linking to somewhere that does?

Author: answered on previous comment. Let me know your thoughts there

## Libp2p Instance Methods 

### start

Starts the libp2p node.

`libp2p.start()`

#### Returns

| Type | Description |
|------|-------------|
| `Promise` | Promise resolves when the node is ready |

#### Example

```js
const Libp2p = require('libp2p')

// ...

const libp2p = await Libp2p.create(options)

// start libp2p
await libp2p.start()
```

### stop

Stops the libp2p node.

`libp2p.stop()`

#### Returns

| Type | Description |
|------|-------------|
| `Promise` | Promise resolves when the node is fully stopped |

#### Example

```js
const Libp2p = require('libp2p')

// ...
const libp2p = await Libp2p.create(options)
// ...

// stop libp2p
await libp2p.stop()
```

### dial

Dials to another peer in the network and establishes the connection.

`dial(peer, options)`

#### Parameters

| Name | Type | Description |
|------|------|-------------|
| peer | [PeerInfo](https://github.com/libp2p/js-peer-info), [PeerId](https://github.com/libp2p/js-peer-id), [multiaddr](https://github.com/multiformats/js-multiaddr), `string` | peer to dial |
| [options] | `Object` | dial options |

> | [options.signal] | `AbortSignal` | abort signal |

Reviewer: ```suggestion
| [options.signal] | [`AbortSignal`](https://developer.mozilla.org/en-US/docs/Web/API/AbortSignal) | An `AbortSignal` instance obtained from an [`AbortController`](https://developer.mozilla.org/en-US/docs/Web/API/AbortController) that can be used to abort the connection before it completes |
```

#### Returns

| Type | Description |
|------|-------------|
| `Promise<Connection>` | Promise resolves with the [Connection](https://github.com/libp2p/js-interfaces/tree/master/src/connection) instance |

#### Example

```js
// ...

> const conn = await libp2p.dial(remotePeerInfo)

Reviewer: Maybe add some more to this example on how someone might use the connection.

```

### dialProtocol

> Dials to another peer in the network and selects a protocol to communicate with that peer.

Reviewer: How does this work? Is it a stream that gets returned not a conneciton?

Author: Yes, I forgot to update the docs here. We return `{ stream, protocol }`

`dialProtocol(peer, protocols, options)`

#### Parameters

| Name | Type | Description |
|------|------|-------------|
| peer | [PeerInfo](https://github.com/libp2p/js-peer-info), [PeerId](https://github.com/libp2p/js-peer-id), [multiaddr](https://github.com/multiformats/js-multiaddr), `string` | peer to dial |

> | protocols | `String|Array<String>` | Strings that identifies the protocols (e.g '/ipfs/bitswap/1.1.0') |

Reviewer: From https://github.com/multiformats/js-multistream-select/#dialerselectprotocols:

```suggestion
| protocols | `String|Array<String>` |  A list of protocols (or single protocol) to negotiate with. Protocols are attempted in order until a match is made. (e.g '/ipfs/bitswap/1.1.0') |
```

| [options] | `Object` | dial options |

> | [options.signal] | `AbortSignal` | abort signal |

Reviewer: ```suggestion
| [options.signal] | [`AbortSignal`](https://developer.mozilla.org/en-US/docs/Web/API/AbortSignal) | An `AbortSignal` instance obtained from an [`AbortController`](https://developer.mozilla.org/en-US/docs/Web/API/AbortController) that can be used to abort the connection before it completes |
```

const conn = await libp2p.dialProtocol(remotePeerInfo, protocols)
```

### hangUp

Closes an open connection with a peer, graciously.

`hangUp(peer)`

#### Parameters

| Name | Type | Description |
|------|------|-------------|
| peer | [PeerInfo](https://github.com/libp2p/js-peer-info), [PeerId](https://github.com/libp2p/js-peer-id), [multiaddr](https://github.com/multiformats/js-multiaddr), `string` | peer to hang up |

#### Returns

| Type | Description |
|------|-------------|
| `Promise<void>` | Promise resolves once connection closes |

#### Example

```js
// ...
await libp2p.hangUp(remotePeerInfo)
```

### handle

Registers a given handler for the given protocols.

`libp2p.handle(protocols, handler)`

#### Parameters

| Name | Type | Description |
|------|------|-------------|
| protocols | `Array<String>|String` | protocols to register |
| handler | `function({ connection:*, stream:*, protocol:string })` | handler to call |

#### Example

```js
// ...
const handler = ({ connection, stream, protocol }) => {
  // use stream or connection according to the needs
}

libp2p.handle('/echo/1.0.0', handler)
```

### unhandle

Unregisters all handlers with the given protocols

> `libp2p.unhandle(protocols)`

Reviewer: What if I only want to unregister a specific handle?

Author: we currently do not support it, but maybe you are right and we should support it. @jacobheun any opinions here for we not supporting it?

Author: I will add a note on `handle`

#### Parameters

| Name | Type | Description |
|------|------|-------------|
| protocols | `Array<String>|String` | protocols to unregister |

#### Example

```js
// ...

> libp2p.unhandle('/echo/1.0.0')

Reviewer 3: We'll convert to an array so I think we should be more specific in examples of our preference, even though we support just passing a string. I'd like to gradually reduce the footprint of accepted params as it increases the likelihood of bugs.
```suggestion
libp2p.unhandle(['/echo/1.0.0'])
```

```

### peerRouting.findPeer

> Iterates over all peer routers in series to find the given peer.

Reviewer 3: ```suggestion
Iterates over all peer routers in series to find the given peer. If the DHT is enabled, it will be tried first.
```

`libp2p.peerRouting.findPeer(peerId, options)`

#### Parameters

| Name | Type | Description |
|------|------|-------------|
| peerId | [`PeerId`](https://github.com/libp2p/js-peer-id) | ID of the peer to find |
| options | `Object` | operation options |
| options.timeout | `number` | maximum time the query should run |

#### Returns

| Type | Description |
|------|-------------|
| `Promise<PeerInfo>` | Peer info of a known peer |

#### Example

```js
// ...
const peerInfo = await libp2p.peerRouting.findPeer(peerId, options)
```

### contentRouting.findProviders

Iterates over all content routers in series to find providers of the given key.
Once a content router succeeds, the iteration will stop.

> `libp2p.contentRouting.findProviders(cid, options)`

Reviewer: Please can this return an async iterable that yields `PeerInfo`.

Reviewer 3: :+1: We haven't refactored this quite yet, should happen in the next day or two, and I intend to make all array based queries async iterable.

Author: Will be changed in the next commit :)

> Once a content router succeeds, the iteration will stop.

Reviewer 3: ```suggestion
Once a content router succeeds, the iteration will stop. If the DHT is enabled, it will be queried first.
```

#### Parameters

| Name | Type | Description |
|------|------|-------------|
| cid | [`CID`](https://github.com/multiformats/js-cid) | cid to find |
| options | `Object` | operation options |
| options.timeout | `number` | maximum time the query should run |
| options.maxNumProviders | `number` | maximum number of providers to find |

#### Returns

| Type | Description |
|------|-------------|
| `Promise<Array<PeerInfo>>` |  array of [`PeerInfo`](https://github.com/libp2p/js-peer-info) |

#### Example

```js
// ...
const providers = await libp2p.contentRouting.findProviders(cid)
```

### contentRouting.provide

Iterates over all content routers in parallel, in order to notify it is a provider of the given key.

`libp2p.contentRouting.provide(cid)`

#### Parameters

| Name | Type | Description |
|------|------|-------------|
| cid | [`CID`](https://github.com/multiformats/js-cid) | cid to provide |

#### Returns

| Type | Description |
|------|-------------|
| `Promise<void>` | Promise resolves once notifications are sent |

#### Example

```js
// ...
await libp2p.contentRouting.provide(cid)
```

> ### _dht.put

Reviewer: Put each of these headings in backticks to avoid problems like this :)

Queries the DHT for a value stored for a given key.

`libp2p._dht.get(key, options)`

#### Parameters

| Name | Type | Description |
|------|------|-------------|
| key | `String` | key to get from the dht |
| [options] | `Object` | get options |
| [options.timeout] | `number` | maximum time the query should run |

#### Returns

| Type | Description |
|------|-------------|
| `Promise<Buffer>` | Value obtained from the DHT |

#### Example

```js
// ...

const key = '/key'
const value = await libp2p._dht.get(key)
```

### _dht.getMany

Queries the DHT for the n values stored for the given key (without sorting).

`libp2p._dht.getMany(key, nvals, options)`

#### Parameters

| Name | Type | Description |
|------|------|-------------|
| key | `String` | key to get from the dht |
| nvals | `number` | number of values aimed |
| [options] | `Object` | get options |
| [options.timeout] | `number` | maximum time the query should run |

#### Returns

| Type | Description |
|------|-------------|
| `Promise<Array<{from: PeerId, val: Buffer}>>` | Array of records obtained from the DHT |

#### Example

```js
// ...

const key = '/key'
const { from, val } = await libp2p._dht.get(key)
```

> ### pubsub.getPeersSubscribed

Reviewer: Commonly being subscribed to something is referred to as being a subscriber. Suggestion: `pubsub.getSubscribers(topic)`?

Author: Oh, I agree with you on this. We already have this on all the `pubsub` repos but I will see what I can do

`libp2p.pubsub.getPeersSubscribed(topic)`

#### Parameters

| Name | Type | Description |
|------|------|-------------|
| topic | `string` | topic to publish |

#### Returns

| Type | Description |
|------|-------------|
| `Array<String>` | peer-id subscribed to the topic |

#### Example

```js
const peerIds = libp2p.pubsub.getPeersSubscribed(topic)
```

### pubsub.getTopics

Gets a list of topics the node is subscribed to.

`libp2p.pubsub.getTopics()`

#### Returns

| Type | Description |
|------|-------------|
| `Array<String>` | topics the node is subscribed to |

#### Example

```js
const topics = libp2p.pubsub.getTopics()
```

### pubsub.publish

Publishes messages to the given topics.

`libp2p.pubsub.publish(topic, data)`

#### Parameters

| Name | Type | Description |
|------|------|-------------|
| topic | `string` | topic to publish |
| data | `Buffer` | data to publish  |

#### Returns

| Type | Description |
|------|-------------|
| `Promise` | publish success |

#### Example

```js
const topic = 'topic'
const data = Buffer.from('data')

await libp2p.pubsub.publish(topic, data)
```

### pubsub.subscribe

Subscribes the given handler to a pubsub topic.

> `libp2p.pubsub.subscribe(topic)`

Reviewer: ```suggestion
`libp2p.pubsub.subscribe(topic, handler)`
```

#### Parameters

| Name | Type | Description |
|------|------|-------------|
| topic | `string` | topic to subscribe |

> | handler | `function(<Object>)` | handler for new data on topic |

Reviewer: Please describe the `Object`!

#### Returns

| Type | Description |
|------|-------------|

> | `void` |  |

Reviewer: Just checking this is not an async operation?

Author: no, we do not need any async operation for any of the pubsub routers in this operation

#### Example

```js
const topic = 'topic'
const handler = (msg) => {
  // msg.data - pubsub data received
}

libp2p.pubsub.subscribe(topic, handler)
```

### pubsub.unsubscribe

Unsubscribes the given handler from a pubsub topic. If no handler is provided, all handlers for the topic are removed.

> `libp2p.pubsub.unsubscribe(topic)`

Reviewer: ```suggestion
`libp2p.pubsub.unsubscribe(topic, handler)`
```
